src/defender_metrics/mod.rs

```diff
@@ -0,0 +1,404 @@
+use crate::shadows::Error;
+use data_types::Metric;
+use embassy_sync::blocking_mutex::raw::RawMutex;
+use embedded_mqtt::{DeferredPayload, Publish, Subscribe, SubscribeTopic, ToPayload};
+use errors::{ErrorResponse, MetricError};
+use futures::StreamExt;
+use serde::Serialize;
+use topics::Topic;
+
+// pub mod aws_types;
+pub mod aws_types;
+pub mod data_types;
+pub mod errors;
+pub mod topics;
+
+pub struct MetricHandler<'a, 'm, M: RawMutex> {
+    mqtt: &'m embedded_mqtt::MqttClient<'a, M>,
+}
+
+impl<'a, 'm, M: RawMutex> MetricHandler<'a, 'm, M> {
+    pub fn new(mqtt: &'m embedded_mqtt::MqttClient<'a, M>) -> Self {
+        Self { mqtt }
+    }
+
+    pub async fn publish_metric<'c, C: Serialize>(
+        &self,
+        metric: Metric<'c, C>,
+        max_payload_size: usize,
+    ) -> Result<(), MetricError> {
+        let payload = DeferredPayload::new(
+            |buf: &mut [u8]| {
+                #[cfg(feature = "metric_cbor")]
+                {
+                    let mut serializer = minicbor_serde::Serializer::new(
+                        minicbor::encode::write::Cursor::new(&mut *buf),
+                    );
+
+                    match metric.serialize(&mut serializer) {
+                        Ok(_) => {}
+                        Err(_) => {
+                            error!("An error happened when serializing metric with cbor");
```

Comment:
This seems like something that should return a serialize error instead of just a log?